### ISSUE TITLE
Distribution on different operating systems

### ISSUE THREAD

**Issue by ppinard** [pymontecarlo/pymontecarlo#26]
The distribution of pymontecarlo to the different operating systems (Windows, MacOS, Linux) is quite complicated due to several factors. 

  - First, it is difficult to freeze (i.e. include the Python distribution) due to the use of setuptools' entry points. These entry points are however essential to allow pymontecarlo to discover installed programs, handlers to parse XML or HDF5 results, GUI interfaces, etc. [py2exe](py2exe.org), [py2app](https://pythonhosted.org/py2app/), [cx_freeze](https://cx-freeze.readthedocs.io/en/latest/) do not support [entry points](http://docs.python-guide.org/en/latest/shipping/freezing/#comparison-of-freezing-tools). For Windows, the project [python_embedded_launcher](https://github.com/zsquareplusc/python-embedded-launcher) works well, but I did not find alternative to MacOS and Linux.

  - Secondly, it is not clear up to now whether pymontecarlo should distribute the Monte Carlo program or ask the user to download/install them separately. I propose that we provide all freely available MC programs with pymontecarlo and configure pymontecarlo accordingly to use them. These would include Casino, WinXRay, PENEPMA, NISTMonte and MONACO. For PENEPMA, I think we should only support the version distributed by the Barcelona group.

  - Thirdly, to facilitate installation, it should be possible to run pymontecarlo directly after extracting it from a ZIP, i.e. no installation, no administrative rights. On Linux, we can probably assume that Python 3 is installed. This assumption should not be made for Windows and MacOS.

**Comment by ppinard** [comment]
@drix00 said:
This one is a big challenge. I need to study it more, but my first question is setuptool entry point is the only solution?

**Comment by ppinard** [comment]
No, and it can probably be replaced by something else. Entry points are stored in a config file in the .egg-info or .dist-info folder. setuptools is only intelligent to know where to find them. For us, we could put everything in a config file and import the entry points from there.

My proposal for the different distributions would be to create a "debug" project where we have numpy, scipy, matplotlib and pyQt as dependencies and try to create distributions for all the platforms.

**Comment by ppinard** [comment]
I created a repository [pymontecarlo_gui_debug](https://github.com/pymontecarlo/pymontecarlo-gui-debug) to test and debug different options to build a pymontecarlo distribution. The GUI interface should look like this:

![image](https://cloud.githubusercontent.com/assets/715280/20829929/31204d54-b876-11e6-9223-7f3589b96759.png)

It requires *numpy*, *scipy*, *h5py*, *PyQt5* and *matplotlib* to be properly installed.

**Comment by ppinard** [comment]
I tested with [cx_freeze](http://cx-freeze.readthedocs.io/en/latest/) version 5.0 under Windows 10 64-bit and I get the following error when starting the frozen program: `Error importing scipy. you cannot import scipy while being in scipy source directory...`

I also wanted to test with [py2exe](http://www.py2exe.org/) and unfortunately these is no version for Python 3.5.

**Comment by drix00** [comment]
For CASINO, Prof Drouin prefer the user download it from the website as he use the download statistic in his finding application and funding report.

Also I did not realize we switch to PyQt from pyside.

To bad for py2exe as I use it for a very simple project and it work well.

I was also thinking with clear instruction and help from already compiled wheel we may not need a frozen application. Maybe look at spyder project for guideline. For our group I make all my project available as wheel and create a batch file to install them. They still need to install the 3rd party packages but I give them clear instruction for that.

**Comment by ppinard** [comment]
There hasn't been a outside release in over a year. There is no Python 3.5 support and no Qt 5. There are wheels for PyQt5 on PyPi.

**Comment by ppinard** [comment]
I was looking at [homebrew](https://github.com/Homebrew/brew) for the distribution on MacOSX. It would require the user to use the command line but it is fairly easy to create "formula" for any application. 

> For CASINO, Prof Drouin prefer the user download it from the website as he use the download statistic in his finding application and funding report.

Are the statistics calculated from the Download form (http://www.gel.usherbrooke.ca/casino/download.html) or the actual ZIP download? In the latter case, a homebrew formula could download CASINO from the website at each user installation and the statistics would still work.

**Comment by drix00** [comment]
> Are the statistics calculated from the Download form (http://www.gel.usherbrooke.ca/casino/download.html) or the actual ZIP download? In the latter case, a homebrew formula could download CASINO from the website at each user installation and the statistics would still work.

The download statistic is calculated from the form when submitted. Maybe we can create a script that fill the form and submit it using httplib or urlib2 module or curl program. We don't need to download the zip file, we can use a local copy.
The 3 input fields and button names are:
`Name email affiliation Submit`

**Comment by ppinard** [comment]
This appears to work:

`curl --form "Name=Test&email=test@pymontecarlo.org&affiliation=pymontecarlo" http://www.gel.usherbrooke.ca/casino/download.php`

**Comment by drix00** [comment]
It only add the name in the file. The & does not work with curl. The correct command is:

`curl --form "Name=Test" --form "email=test@pymontecarlo.org" --form "affiliation=pymontecarlo" http://www.gel.usherbrooke.ca/casino/download.php`

I verify and the entry is added in the file used to count the number of download with this command.

**Comment by ppinard** [comment]
I created homebrew "formulae" for Casino 2, WinX-Ray and [pymontecarlo-gui-debug](https://github.com/pymontecarlo/pymontecarlo-gui-debug), which seem to work. They are in the project [homebrew-pymontecarlo](https://github.com/pymontecarlo/homebrew-pymontecarlo). The one for Casino 2 calls the above curl function after the installation. However, only the name is sent as the other fields cannot be found from terminal commands. I tried to have an user input, but homebrew does not seem to allow this.

**Comment by drix00** [comment]
It does not matter, Casino team does not contact the user via that information. The main goal is to have an idea of the number of download. Maybe add a default email of the project pymontecarlo if it exist and add in the affiliation to `automatically download by pyMonteCarlo on MacOS`. That way I can track the number of download from pyMonteCarlo. If curl is also used for Windows and Linux, maybe add it also.

**Comment by ppinard** [comment]
I followed the instruction to create a [embedded Python distribution](https://docs.python.org/3.5/using/windows.html#embedded-distribution) on Windows. It seems to work well, but it requires to have Visual Studio installed to compile a `.exe`. Could someone test it out? 

1. Download the [ZIP from dropbox](https://www.dropbox.com/s/jnstfvyusppi8up/pyMonteCarlo-GUI-debug-0.1.2%2B0.g413058e.dirty.zip?dl=0) (220Mb)
2. Extract files
3. Run pymontecarlo-gui-debug.exe
4. After a few seconds of loading, you should see something similar to [this](https://github.com/pymontecarlo/pymontecarlo/issues/26#issuecomment-264417555)

**Comment by drix00** [comment]
It work on Windows 8.1, but I have visual studio already installed. The log and h5 files are created in the user directory under `.pymontecarlo` folder.

I will try on different version of windows tomorrow, I have in virtualbox of window 2000, XP, 7, 8.1, and 10 on the university workstation. We need to check if we need to have Microsoft C Runtime to run pymontecarlo-gui-debug.exe. 

Could you add the steps needed to create the zip and pymontecarlo-gui-debug.exe.

Also we may try to use NSIS (http://nsis.sourceforge.net/) or WiX (http://wixtoolset.org/) to create and installer (and download/install the Microsoft C Runtime if needed).

**Comment by ppinard** [comment]
It would be great to know if it works on Windows 7 and others. I personally tested it on Win 10. We do need to have the Microsoft C Runtime installed (see [note](https://docs.python.org/3.5/using/windows.html#embedded-distribution) in the Python doc). 

The code to build the Windows distribution is for the moment in the [setup.py](https://github.com/pymontecarlo/pymontecarlo-gui-debug/blob/embed/setup.py) of pymontecarlo-gui-debug (branch: embed). In short, the command:

1. Builds the wheel of the project (i.e. pymontecarlo-gui-debug)
2. Downloads the Python Embedded distribution (currently forced to Python 3.5.2)
3. Installs pip
4. Installs pymontecarlo-gui-debug using pip. The dependencies are automatically found and either downloaded from the internet or a folder containing wheels can be specified. This is required for numpy, scipy and h5py as they do not have wheels for Windows on PyPi.
5. Create a small C program to launch Python with the command to start the GUI (see `PY_MAIN_EXE` class variable)
6. Compile the program to create an EXE

Is it really necessary to create an installer? I was thinking that it is easier just to extract files from a ZIP. For the Microsoft Runtime, we can have a warning message on the download page.

**Comment by drix00** [comment]
More systems I try on:

- XP 32-bit, but I got an error message that `pymontecarlo-gui-debug.exe` is not a valid 32-bit application, which is OK as `pymontecarlo-gui-debug.exe`  is a 64-bit executable.
- Work on 2 windows 7 system, but I did not have to install the Microsoft C Runtime.
- Problem with another window 8.1 system, the Microsoft C Runtime was not able to install. I will do the update and see if I can install the Microsoft C Runtime.

So I was not able to test when I don't have the Microsoft C Runtime already install.

> Is it really necessary to create an installer? I was thinking that it is easier just to extract files from a ZIP. For the Microsoft Runtime, we can have a warning message on the download page.

We could at least include the correct Microsoft C Runtime `vc_redist.x64.exe` in the zip file as they need to download the one for the correct Visual studio (2015 for python 3.5+). 
https://www.microsoft.com/en-us/download/details.aspx?id=53587

> After a few seconds of loading, you should see something similar to this

Was longer on window 7, more like a minute to start it.

**Comment by ppinard** [comment]
> We could at least include the correct Microsoft C Runtime vc_redist.x64.exe in the zip file as they need to download the one for the correct Visual studio (2015 for python 3.5+).
https://www.microsoft.com/en-us/download/details.aspx?id=53587

Good idea. I will add this to the script.

**Comment by ppinard** [comment]
I think we might be able to close the issue and transfer the decisions and processes to the wiki.

- Windows: Python embedded distribution, pre-installed pymontecarlo using wheels and vc_redist_x64.exe. We also include supported Monte Carlo codes in the zip. 

- MacOS: homebrew formula for pymontecarlo and separate ones Monte Carlo codes. We can perhaps have a `pymontecarlo-all` formula that installs everything.

- Linux: I started making some Debian package (.deb) for the Monte Carlo code (see https://packagecloud.io/ppinard/pymontecarlo). Maybe we need a deb for pymontecarlo or maybe simply use pip.

**Comment by drix00** [comment]
Some other options to look for the future.

On windows, I have use WinPython lately to use Python script without installing python on the computer. It is similar to the Python embedded distribution, but you can also modify and run the code using Spyder. This allow a more advance workshop by including scripting pymontecarlo. You still need to install the correct  Microsoft C Runtime.

The anaconda distribution seem to be popular with scientific packages and the conda is package management which is suppose to be more power full than pip. But I never try it.

**Comment by ppinard** [comment]
Does WinPython use pip?

I tried conda and the problem was that it is not really portable. If you install conda in one directory, you cannot rename or move that directory.

**Comment by drix00** [comment]
> Does WinPython use pip?

Yes, they have a package manager, but you can use pip. For my current project, I have to use python 2.7 and they only have the barebone distribution available (called Zero Version), but I was able to install the package from pypi or http://www.lfd.uci.edu/~gohlke/pythonlibs/. 

My workflow is to develop my project on my laptop with my normal python distribution. Copy or use git pull on the external hard drive with WinPython, first test it on my laptop but with WinPython. Than connect the external hard drive to the instrument computer and use the project and sometime modify the code on the instrument computer. Making the commit later on my laptop.

I have made copy of the scripts and WinPython on another external hard drive and run them without problem. The only problem I have found is if you install your package in developer mode (pip install -e .), the Windows drive letter change between computer, so you have to install it each time you change computer. But if you do a normal install it will be fine.

> I tried conda and the problem was that it is not really portable. If you install conda in one directory, you cannot rename or move that directory.

For conda, I was not thinking about a portable solution like WinPython, but a solution to include external program like penelope and casino in the installation. But again I did not try it so I don't know if it is possible.

**Comment by ppinard** [comment]
Looking back at how the installation of pyMonteCarlo went during the EMAS workshop, I think the installation on MacOS has to be simplified. We need to find a way to create a *pyMonteCarlo.app*. The Windows installation as it stands right now worked perfectly.
